src/main/java/io/odpf/firehose/sinkdecorator/SinkWithDlq.java

```diff
@@ -5,97 +5,111 @@
  */
 public class SinkWithDlq extends SinkDecorator {
 
-    private Producer<byte[], byte[]> kafkaProducer;
-    private final String topic;
-    private BackOffProvider backOffProvider;
+    public static final String DLQ_BATCH_KEY = "dlq-batch-key";
+    private final DlqWriter writer;
+    private final BackOffProvider backOffProvider;
+    private final int maxRetryAttempts;
+    private final boolean isFailOnMaxRetryAttemptsExceeded;
+
     private Instrumentation instrumentation;
 
-    public SinkWithDlq(Sink sink, Producer<byte[], byte[]> kafkaProducer, String topic,
-                       Instrumentation instrumentation, BackOffProvider backOffProvider) {
+    public SinkWithDlq(Sink sink, DlqWriter writer, BackOffProvider backOffProvider, int maxRetryAttempts, boolean isFailOnMaxRetryAttemptsExceeded, Instrumentation instrumentation) {
         super(sink);
-        this.kafkaProducer = kafkaProducer;
-        this.topic = topic;
+        this.writer = writer;
         this.backOffProvider = backOffProvider;
+        this.maxRetryAttempts = maxRetryAttempts;
+        this.isFailOnMaxRetryAttemptsExceeded = isFailOnMaxRetryAttemptsExceeded;
         this.instrumentation = instrumentation;
     }
 
-    public static SinkWithDlq withInstrumentationFactory(Sink sink, Producer<byte[], byte[]> kafkaProducer, String topic,
-                                                         StatsDReporter statsDReporter, BackOffProvider backOffProvider) {
-        return new SinkWithDlq(sink, kafkaProducer, topic, new Instrumentation(statsDReporter, SinkWithDlq.class), backOffProvider);
+    public static SinkWithDlq withInstrumentationFactory(Sink sink, DlqWriter dlqWriter, BackOffProvider backOffProvider, int maxRetryAttempts, boolean isFailOnMaxRetryAttemptsExceeded, StatsDReporter statsDReporter) {
+        return new SinkWithDlq(sink, dlqWriter, backOffProvider, maxRetryAttempts, isFailOnMaxRetryAttemptsExceeded, new Instrumentation(statsDReporter, SinkWithDlq.class));
     }
 
     /**
      * Pushes all the failed messages to kafka as DLQ.
      *
-     * @param message list of messages to push
+     * @param inputMessages list of messages to push
      * @return list of failed messaged
      * @throws IOException
      * @throws DeserializerException
      */
     @Override
-    public List<Message> pushMessage(List<Message> message) throws IOException, DeserializerException {
-        List<Message> retryQueueMessages = super.pushMessage(message);
-        int attemptCount = 0;
+    public List<Message> pushMessage(List<Message> inputMessages) throws IOException, DeserializerException {
+        List<Message> dlqMessages = super.pushMessage(inputMessages);
 
-        while (!retryQueueMessages.isEmpty()) {
-            instrumentation.captureRetryAttempts();
-            retryQueueMessages = pushToKafka(retryQueueMessages);
-            backOffProvider.backOff(attemptCount++);
+        List<Message> dlqWriteFailedMessages;
```

Comment:
we can simplify with this.
`
        List<Message> dlqWriteFailedMessages;
        dlqWriteFailedMessages = pushToWriter(dlqMessages);
        LinkedList<Message> dlqProcessedMessages = relativeComplementSet(dlqMessages, dlqWriteFailedMessages);
        if (super.canManageOffsets()) {
            super.addOffsets(DLQ_BATCH_KEY, dlqProcessedMessages);
            super.setCommittable(DLQ_BATCH_KEY);
        }
        instrumentation.logInfo("DLQ processed messages: {}", dlqProcessedMessages.size());
`

Reply:
we can add offsets at the end. It's fine because it's the single thread. Also in the case of sink not managing offsets, it's happening after pushMessage is finished. WE don't have to be so active about adding offsets.

Reply:
if we want we can add offsets before throwing Exception just be be complete. 
But from the current behaviour: firehose consumer fails if pushmessage() throws exception. so adding offsets doesn't matter.

Reply:
will simplify this